xdsl/transforms/convert_snrt_to_riscv.py

```diff
@@ -395,10 +411,141 @@ def match_and_rewrite(
+@dataclass
+class LowerClusterCoreNum(RewritePattern):
+    constants: SnrtConstants
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: snitch_runtime.ClusterCoreNumOp, rewriter: PatternRewriter, /
+    ):
+        rewriter.replace_matched_op(
+            [
+                nfo := riscv.LiOp(self.constants.SNRT_CLUSTER_CORE_NUM),
+                builtin.UnrealizedConversionCastOp.get([nfo], [builtin.i32]),
+            ]
+        )
+
+
+@dataclass
+class LowerClusterNum(RewritePattern):
+    constants: SnrtConstants
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: snitch_runtime.ClusterNumOp, rewriter: PatternRewriter, /
+    ):
+        rewriter.replace_matched_op(
+            [
+                nfo := riscv.LiOp(self.constants.SNRT_CLUSTER_NUM),
+                builtin.UnrealizedConversionCastOp.get([nfo], [builtin.i32]),
+            ]
+        )
+
+
+@dataclass
+class LowerClusterDmCoreNum(RewritePattern):
+    constants: SnrtConstants
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: snitch_runtime.ClusterDmCoreNumOp, rewriter: PatternRewriter, /
+    ):
+        rewriter.replace_matched_op(
+            [
+                nfo := riscv.LiOp(self.constants.SNRT_CLUSTER_DM_CORE_NUM),
+                builtin.UnrealizedConversionCastOp.get([nfo], [builtin.i32]),
+            ]
+        )
+
+
+@dataclass
+class LowerClusterComputeCoreNum(RewritePattern):
+    constants: SnrtConstants
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: snitch_runtime.ClusterComputeCoreNumOp, rewriter: PatternRewriter, /
+    ):
+        """
+        Implementation:
+
+        inline uint32_t __attribute__((const)) snrt_cluster_compute_core_num() {
+            return snrt_cluster_core_num() - snrt_cluster_dm_core_num();
+        }
+
+        inline uint32_t __attribute__((const)) snrt_cluster_core_num() {
+            return SNRT_CLUSTER_CORE_NUM;
+        }
+
+        inline uint32_t __attribute__((const)) snrt_cluster_dm_core_num() {
+            return SNRT_CLUSTER_DM_CORE_NUM;
+        }
+        """
+        rewriter.replace_matched_op(
+            [
+                compute_core_num := riscv.LiOp(
+                    self.constants.SNRT_CLUSTER_CORE_NUM
+                    - self.constants.SNRT_CLUSTER_DM_CORE_NUM
+                ),
+                builtin.UnrealizedConversionCastOp.get(
+                    [compute_core_num], [builtin.i32]
+                ),
+            ]
+        )
+
+
+@dataclass(frozen=True)
 class ConvertSnrtToRISCV(ModulePass):
+    """
+    Convert snitch runtime operations to their definitions as per the snitch runtime spec.
+    """
+
     name = "convert-snrt-to-riscv"
 
+    cluster_num: int = 4
+    cluster_core_num: int = 9
+    base_hartid: int = 0
+    cluster_dm_core_num: int = 1
+
     def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        constants = SnrtConstants(self.cluster_num)
```

Comment:
```suggestion
        constants = SnrtConstants(self.cluster_num, self.cluster_core_num, self.base_hartid, self.cluster_dm_core_num)
```

Reply:
Fixed this by making the pass inherit from the `SnrtConstants` dataclass, it now integrates very nicely.